api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters/scenario-data/derive-scenario.facade.ts

```diff
@@ -1,11 +1,48 @@
-import { Injectable } from '@nestjs/common';
-import { InputFiles } from '../../ports/input-files';
+import { resolve, dirname } from 'path';
+import { createWriteStream, promises } from 'fs';
+import { HttpService, Injectable } from '@nestjs/common';
+import { Assets, InputFiles } from '../../ports/input-files';
 
 @Injectable()
 export class DeriveScenarioFacade implements InputFiles {
-  async include(scenarioId: string, directory: string): Promise<void> {
-    // get all scenario data from db
-    // and transform them into files required by marxan
+  constructor(private readonly httpService: HttpService) {}
+
+  async include(directory: string, assets: Assets): Promise<void> {
+    // TODO security: ensure that none of the target filename starts with either `/` or `.` (tree traversing)
+    await Promise.all(
+      assets.map((asset) =>
+        this.download(asset.url, resolve(directory, asset.relativeDestination)),
+      ),
+    );
     return;
   }
+
+  private async download(sourceUri: string, dest: string) {
+    await this.ensureWriteDirectoryExists(dest);
+    return new Promise((resolve, reject) => {
+      const writer = createWriteStream(dest);
+      writer.on('finish', resolve);
+      writer.on('error', reject);
+      // "stream" seems to be broken, does not return stream but data itself
+      // https://github.com/nestjs/nest/issues/4144 - not really working
+      this.httpService
+        .get(sourceUri, {
+          responseType: 'stream',
+        })
+        .toPromise()
+        .then((response) => {
+          writer.write(response.data);
+          writer.end();
+        })
+        .catch(reject);
+    });
```

Comment:
hmm, so if `stream` does not work and it is not really streaming, why do you use `createWriteStream`? I'd suggest to just use `writeFile`

Reply:
Thought maybe someone can pinpoint some obvious thing I missed. :(